### PR TITLE
feat(collections): add dropWhile

### PR DESCRIPTION
**Pull request by getspooky**
Related: https://github.com/denoland/deno_std/issues/1173

### CODE REVIEW

**Comment by LionC** [comment]
Would you mind changing the title to a simple `feat(collections): add dropWhile`? That would help with overview, as there is lots of small PRs adding functions.

I also think this function should be named `dropWhile` to match up with `takeWhile` :-)

**Comment by getspooky** [comment]
@LionC sure i will implement this changes , I think something went wrong with `CI/CD`

**Comment by LionC** [comment]
`dropWhile` vs `dropLastWhile` is in line with `takeWhile` and `takeLastWhile`, which is in line with the current ES iterator proposal naming for `take` and with kotlin, which set most of the other names.

**Comment by getspooky** [comment]
@LionC could you please review this PR

File: collections/README.md [hunk 1]
```diff
@@ -112,6 +112,20 @@ const distinctNumbers = distinct(numbers);
 console.assert(distinctNumbers === [3, 2, 5]);
 ```
 
+### dropWhile
+
+Drop all elements in the given collection until the first element that does not
+match the given predicate.
```
Comment (LionC) [inline collections/README.md:118]: This should make clear that it does not touch the original array. We should also not use the word "collection" - it just takes an Array right now :-)

File: collections/drop_while.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Drop all elements in the given collection until the first element that does not match the given predicate.
+ *
```
Comment (LionC) [inline collections/drop_while.ts:5]: Same here obv

File: collections/drop_while.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * Example:
+ *
+ * ```ts
+ * import { dropWhile } from "./drop_first_while.ts";
+ * import { assertEquals } from "../testing/asserts.ts";
+ *
+ * const numbers = [ 3, 2, 5, 2, 5 ]
+ * const dropWhileNumbers = dropWhile(numbers, i => i !== 2)
+ *
+ * assertEquals(dropWhileNumbers, [ 5, 2, 5 ])
```
Comment (LionC) [inline collections/drop_while.ts:15]: That examples looks wrong. Shouldn't this be `[ 2, 5, 2, 5 ]`?

File: collections/drop_while.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  predicate: (el: T) => boolean,
+): T[] {
+  let offset = 0;
+  const length = array.length;
+
+  while (length > offset && predicate(array[offset])) {
+    offset++;
+  }
+
+  return array.slice(offset + 1, length);
```
Comment (LionC) [inline collections/drop_while.ts:29]: I think the `+ 1` is actually a bug isnt it?

I would expect `dropWhile` to drop everything that fulfils the predicate. This actually just drops one more element.

File: collections/drop_while_test.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  assertEquals(actual, []);
+});
+
+Deno.test("[collections/dropWhile] Returns the same array when all elements match the predicate", () => {
+  const arr = [1, 2, 3, 4];
+
+  const actual = dropWhile(arr, (i) => i !== 400);
+
+  assertEquals(actual, []);
+});
```
Comment (LionC) [inline collections/drop_while_test.ts:57]: The tests seem to test the buggy behaviour mentioned above.

**Review by LionC — CHANGES_REQUESTED** [review]
I think the current implementation does not behave as expected -- see the comments

File: collections/drop_while.ts (hunk 3 above)
Comment (getspooky) [reply collections/drop_while.ts:15]: As i know we should not include predictable number,

Comment (LionC) [reply collections/drop_while.ts:15]: I don't think I understand what you mean.

`dropWhile` should drop all elements from the start that match the given predicate, until it hits one that doesnt. The one that doesnt should not be dropped, as it does not match the predicate.

File: collections/drop_while.ts (hunk 4 above)
Comment (getspooky) [reply collections/drop_while.ts]: fixed

File: collections/drop_while.ts (hunk 3 above)
Comment (getspooky) [reply collections/drop_while.ts]: @LionC fixed

**Review by LionC — APPROVED** [review]
LGTM! Pinging @kt3k

**Review by kt3k — APPROVED** [review]
@getspooky LGTM. Thanks!

**Pull request merged**